vcd/resource_vcd_vapp_vm.go

```diff
@@ -287,20 +324,32 @@ func resourceVcdVAppVmCreate(d *schema.ResourceData, meta interface{}) error {
 	}
 
 	log.Printf("[TRACE] Creating VM: %s", d.Get("name").(string))
-	task, err := vapp.AddNewVM(d.Get("name").(string), vappTemplate, &networkConnectionSection, acceptEulas)
+	var storageProfile types.Reference
+	storageProfilePtr := &storageProfile
+	storageProfileName := d.Get("storage_profile").(string)
+	if storageProfileName != "" {
+		storageProfile, err = vdc.FindStorageProfileReference(storageProfileName)
+		if err != nil {
+			return fmt.Errorf("[vm creation] error retrieving storage profile %s : %s", storageProfileName, err)
+		}
+	} else {
+		storageProfilePtr = nil
+	}
+	task, err := vapp.AddNewVMWithStorageProfile(d.Get("name").(string), vappTemplate, &networkConnectionSection, storageProfilePtr, acceptEulas)
```

Comment:
Doesn't this work by avoiding  storageProfilePtr and just having var and passing `*storageProfile` ?

Reply:
No. We need to pass a `nil` when there is no storage profile. 
If we pass `*storageProfile` it would be a pointer to an empty structure